### ISSUE TITLE
Interactive shell is very slow

### ISSUE THREAD

**Issue by lostinroom** [phpseclib/phpseclib#1178]
Hi, 

I make slightly modification of http://phpseclib.sourceforge.net/ssh/examples.html.

So I have this code:
echo $ssh->read('username@username:~$');
$ssh->write("whoami; date;\n");
echo $ssh->read('username@username:~$');
$ssh->write("pwd; date;\n");
echo $ssh->read('username@username:~$');

It takes around 10 secs from first command (whoami) to second command (pwd).
Why does it happen?
Can it be improved?

Thanks.

**Comment by terrafrost** [comment]
Odds are that the string you're using to identify your prompt are incorrect.

Maybe there's ANSI in the prompt that you're not accounting for. If your prompt (for example) has colors in it then that's probably ANSI.

Maybe you're actually using `username@username:$` as your prompt instead of realizing that it's a placeholder.

In order to know for sure I'd need to see two things: your PHP code and your SSH logs. You can get your SSH logs by doing `define('NET_SSH2_LOGGING', 2);` at the top and `echo $ssh->getLog();` after your first prompt.

As for why it'd be taking so long when an invalid prompt is used...  it's waiting for 10s until it does see the prompt and then returns everything that it did see when it does get the prompt. I suppose one thing you could do is lower the timeout (eg. `$ssh->setTimeout(5)`) but the best solution is to just make it use the correct string for the prompt.

**Comment by lostinroom** [comment]
Hi, thanks for quick response.
I lower the timeout as you suggest it, yes, it speeds up.
Why is that? Is there any problem if change timeout to 1?

This is my php code:
require_once 'vendor/autoload.php';

$loader = new \Composer\Autoload\ClassLoader();
$loader->addPsr4('phpseclib\\', __DIR__.'/phpseclib');
$loader->register();

use phpseclib\Net\SSH2;

define('NET_SSH2_LOGGING', 2);

try {
	$ssh = new SSH2('spreod03');
	$ssh->setTimeout(5);
	
	if (!$ssh->login('user001', 'xxxx')) {
		throw new Exception('Login Failed');
	}
	
	echo '<pre>';
	echo $ssh->read('username@username:~$');
	$ssh->write("whoami;date;\n");
	echo $ssh->read('username@username:~$');
	$ssh->write("pwd;date;\n");
	echo $ssh->read('username@username:~$');

	var_dump($ssh->getLog());
	echo '</pre>';
} catch(Exception $e) {
	echo '<pre>';
	var_dump($e->getMessage());
	echo '</pre>';
}

SSH2 logs:
-> NET_SSH2_MSG_CHANNEL_DATA (since last: 2.5933, network: 0.0001s)
00000000  00:00:00:00:00:00:00:0d:77:68:6f:61:6d:69:3b:64  ........whoami;d
00000010  61:74:65:3b:0a                                   ate;.

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0008, network: 0s)
00000000  00:00:00:01:00:00:00:04:77:68:6f:61              ........whoa

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:02:6d:69                    ........mi

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:06:3b:64:61:74:65:3b        ........;date;

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:03:0d:0d:0a                 ...........

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.3209, network: 0.0001s)
00000000  00:00:00:01:00:00:00:09:63:70:72:61:6e:6f:74:0d  ........user001.
00000010  0a                                               .

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0029, network: 0s)
00000000  00:00:00:01:00:00:00:1e:54:68:75:20:41:75:67:20  ........Thu Aug 
00000010  33:31:20:30:39:3a:31:34:3a:33:33:20:53:47:54:20  31 09:14:33 SGT 
00000020  32:30:31:37:0d:0a                                2017..

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0004, network: 0s)
00000000  00:00:00:01:00:00:00:0d:56:31:5f:73:70:72:65:6f  ........V1_spreo
00000010  64:30:33:25:20                                   d03% 

-> NET_SSH2_MSG_CHANNEL_DATA (since last: 4.6805, network: 0.0001s)
00000000  00:00:00:00:00:00:00:0a:70:77:64:3b:64:61:74:65  ........pwd;date
00000010  3b:0a                                            ;.

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0007, network: 0s)
00000000  00:00:00:01:00:00:00:01:70                       ........p

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:02:77:64                    ........wd

<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0001, network: 0s)
00000000  00:00:00:01:00:00:00:06:3b:64:61:74:65:3b        ........;date;

**Comment by lostinroom** [comment]
I just realized that my bash prompt is different. 
It is V1_[my machine name]% not username@username:~$.
How do I know what my default bash prompt?

**Comment by terrafrost** [comment]
Your prompt appears to be `V1_spreod03%`. I'm gathering that from this packet:

```
<- NET_SSH2_MSG_CHANNEL_DATA (since last: 0.0004, network: 0s)
00000000  00:00:00:01:00:00:00:0d:56:31:5f:73:70:72:65:6f  ........V1_spreo
00000010  64:30:33:25:20                                   d03% 
```

The way I usually figure out is...  I usually do `$ssh->setTimeout(3); echo $ssh->read();`. It'll usually be the last line in the output.

If you have a bunch of computers with the `V1_spreod03%` prompt convention you could do a regular expression like `#V1_\w+%#`, with the caveat that the more generic you make the parameter the more likely you are to encounter false positives. Like if you did `tail -f filename.ext` on a file that had `V1_spread03%` at the top it'd stop on that and not the actual prompt.

As for using `setTimeout(1)`...  you can try it out but in my experience a sufficiently short timeout is liable to time out before you actually get the prompt. `setTimeout(0)` means that there is no timeout - that requests will block until either PHP times out, the connection times out or it sees the expected output.

> How do I know what my default bash prompt?

That's more of a Linux question than it is an SSH / phpseclib question. https://linuxconfig.org/bash-prompt-basics provides a good discussion however.

**Comment by lostinroom** [comment]
Hi, thanks for the explanation. 
After doing some testing my end, I realized that I can't find pattern of my prompt.
I have around 20 machines and they are set by different person so they may have different preference to set up the prompt format.
i.e.  V1_spreod03% and spr004sg:~%

Is there any other way without providing my prompt format to $ssh->read()?
From my opinion, it may cause issue, if users personalize their prompt format.

**Comment by terrafrost** [comment]
Not really. Some of the problem is that `$ssh->read()` doesn't have to be used in the context of a shell. You could use it with vim, for example. You send "i" to insert text and then "asdf" to insert the string "asdf" into the file you're editing and then you send "[esc]". Then you send ":w\n". At this point you may want to wait for vim to say whether or not it saved the file or some such and conditionally branch off from there.

In an interactive shell Linux can send you data even tho you're in a prompt. "_The system is going down for system halt NOW!_" is probably the best example of this but, more generally, there's [wall](https://en.wikipedia.org/wiki/Wall_(Unix)) that can do it as well. Just because you legit encounter the prompt (and by legit encounter I mean it's not being printed out via `cat` or some such) doesn't mean that the server won't continue to send data.

Overall, `read()` provides a lot of versatility - probably more than most people ever use. Unfortunately, I'm not really sure what the alternative is.

I guess the best bet for you might be to simply do `setTimeout(3)` or some such. You could try lower timeout's but, in my experience, 3 seems to the best length.
